// generator-prismic-nextjs/generators/slicemachine/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const prismic_yeoman_generator_1 = require("@prismicio/prismic-yeoman-generator");
const chalk = require("chalk");
const fs_1 = require("fs");
const { SM_FILE } = require('sm-commons/consts');
class SliceMachine extends prismic_yeoman_generator_1.default {
    constructor(argv, opts) {
        super(argv, opts);
        if (this.destinationRoot().endsWith(this.path) === false) {
            this.destinationRoot(this.path);
        }
        if (opts.framework) {
            this.config.set('framework', opts.framework);
            this.framework = opts.framework;
        }
        else {
            this.framework = this.config.get('framework') || 'nextjs';
            this.config.set('framework', this.framework);
        }
        if (opts.domain) {
            this.domain = opts.domain;
        }
    }
    async prompting() {
        const base = new URL(this.prismic.base);
        if (!this.domain) {
            const validateRepositoryName = (domain) => this.prismic.validateRepositoryName(domain, this.existingRepo);
            const domainPrompt = {
                type: 'input',
                name: 'domain',
                message: 'Name your prismic repository: ',
                transformer(value) {
                    const reponame = value ? chalk.cyan(value) : chalk.dim.cyan('repo-name');
                    const msg = [
                        chalk.dim(`${base.protocol}//`),
                        reponame,
                        chalk.dim(`.${base.hostname}`),
                    ];
                    return msg.join('');
                },
                validate: async (value) => {
                    const result = await validateRepositoryName(value);
                    return result === value || result;
                },
            };
            this.domain = await this.prompt(domainPrompt).then(res => res.domain);
        }
        if (!this.pm)
            await this.promptForPackageManager();
    }
    async writing() {
        const pkgJson = {
            scripts: {
                slicemachine: 'start-slicemachine --port 9999',
            },
            dependencies: {
                'next-slicezone': '^0.0.14',
                'next-transpile-modules': '^6.4.0',
                'theme-ui': '^0.6.2',
                'essential-slices': '^1.0.3',
                '@prismicio/client': '^5',
                'prismic-reactjs': '^1.3.3',
            },
            devDependencies: {
                '@babel/core': '^7.12.10',
                'slice-machine-ui': '^0.1.2',
            },
        };
        this.fs.extendJSON(this.destinationPath('package.json'), pkgJson);
        // theses files could be removed from this package but would have to come from create-next-app
        const useSrc = fs_1.existsSync(this.destinationPath('src'));
        this.fs.copyTpl(this.templatePath(), this.destinationPath(), {
            smFile: SM_FILE,
            useSrc,
            defaultLibrary: 'essential-slices',
            domain: this.domain,
        }, undefined, {
            globOptions: { dot: true },
        });
        if (useSrc) {
            this.moveDestination('pages', 'src');
        }
        // maybe rename sm file
        if (this.existsDestination('sm.json') && SM_FILE !== 'sm.json') {
            this.moveDestination('sm.json', SM_FILE);
        }
        const customTypes = this.readCustomTypesFrom('customtypes');
        return this.maybeCreatePrismicRepository({ domain: this.domain, framework: 'next', customTypes }, this.existingRepo);
    }
    async install() {
        if (this.pm === 'yarn') {
            this.yarnInstall();
        }
        else {
            this.npmInstall();
        }
    }
    async end() {
        const url = new URL(this.prismic.base);
        if (this.domain) {
            url.hostname = `${this.domain}.${url.hostname}`;
            url.pathname = 'documents';
        }
        else {
            url.pathname = 'dashboard';
        }
        const writingRoomUrl = url.toString();
        this.log(`
Your project is now configured to use SliceMachine!
Follow these next steps to get going:
    
- Add the SliceZone, anywhere in your code https://github.com/prismicio/slice-machine/tree/master/packages/next-slicezone
    
- Access your Prismic writing room here
${writingRoomUrl}
    
- To add your own slice, run this command
$> npx prismic-cli sm --create-slice
    
- Run slicemachine
$> npx prismic-cli sm --develop
    
If you're not sure where you should start,
give SliceMachine documentation a try:
    
- Quick Start 👉 https://prismic.io/docs/technologies/quick-start-nextjs
- Adding slices 👉 https://prismic.io/docs/technologies/create-your-own-slices-components-nextjs
`);
    }
}
exports.default = SliceMachine;
