// prismic-cli/lib/commands/create-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
const yeoman_environment_1 = require("yeoman-environment");
class CreateGenerator extends command_1.Command {
    async run() {
        const _a = this.parse(CreateGenerator).flags, { folder } = _a, flags = tslib_1.__rest(_a, ["folder"]);
        const env = yeoman_environment_1.createEnv();
        env.register(require.resolve('@prismicio/prismic-generator-generator'), 'generator');
        // @ts-expect-error
        return env.run('generator', Object.assign(Object.assign({}, flags), { path: folder }));
    }
}
exports.default = CreateGenerator;
CreateGenerator.description = 'Create a new generator.';
CreateGenerator.hidden = true;
CreateGenerator.flags = {
    help: command_1.flags.help({ char: 'h' }),
    name: command_1.flags.string({ char: 'n', description: 'Name for the generator.' }),
    force: command_1.flags.boolean({ description: 'Overwrite local files.' }),
    'skip-install': command_1.flags.boolean({ description: 'prevent packages from being installed' }),
    pm: command_1.flags.string({
        char: 'p',
        description: 'Package manager.',
        required: false,
        options: ['npm', 'yarn'],
    }),
    language: command_1.flags.string({
        char: 'l',
        description: 'Language to write the generator in.',
        required: false,
        options: ['js', 'ts'],
        parse: value => {
            switch (value) {
                case 'js': return 'javascript';
                case 'ts': return 'typescript';
                default: return value;
            }
        },
    }),
    folder: command_1.flags.string({
        char: 'f',
        description: 'Project where to create the generator.',
    }),
};
