// prismic-cli/lib/commands/login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const prismic_1 = require("../prismic");
const server_1 = require("../utils/server");
class Login extends prismic_1.Command {
    async run() {
        const { flags } = this.parse(Login);
        const { base, port, 'auth-url': authUrl } = flags;
        return this.login(port, base, authUrl);
    }
}
exports.default = Login;
Login.description = 'Log in to Prismic.';
Login.flags = Object.assign(Object.assign({}, prismic_1.Command.flags), { help: command_1.flags.help({ char: 'h' }), base: command_1.flags.string({
        name: 'base',
        hidden: true,
        description: 'Base URL to authenticate with.',
        default: 'https://prismic.io',
    }), port: command_1.flags.integer({
        name: 'port',
        hidden: false,
        description: 'Port to start the local login server.',
        default: server_1.DEFAULT_PORT,
    }), 'auth-url': command_1.flags.string({
        name: 'auth url',
        hidden: true,
        description: 'URL to validate and refresh tokens.',
        required: false,
    }) });
Login.args = [];
