// prismic-cli/lib/commands/new.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const inquirer = require("inquirer");
const prismic_1 = require("../prismic");
const yeoman_env_1 = require("../prismic/yeoman-env");
class New extends prismic_1.Command {
    async run() {
        const isAuthenticated = await this.prismic.isAuthenticated();
        if (!isAuthenticated) {
            await this.login();
        }
        const { flags } = this.parse(New);
        const existingRepo = flags['existing-repo'] || false;
        const domain = await this.validateDomain(flags.domain, existingRepo);
        const folder = await this.validateFolder(flags.folder, domain, flags.force);
        const generators = yeoman_env_1.names.map(value => {
            const nameWithOutPrefix = value.replace('prismic-', '').replace(/js$/i, 'JS');
            return {
                name: nameWithOutPrefix.charAt(0).toUpperCase() + nameWithOutPrefix.slice(1),
                value,
            };
        });
        const isValidTemplate = flags.template && yeoman_env_1.names.includes(`prismic-${flags.template.toLowerCase()}`) ? `prismic-${flags.template.toLowerCase()}` : '';
        const template = isValidTemplate || await inquirer.prompt({
            type: 'list',
            name: 'template',
            message: 'Template to use',
            choices: generators,
        }).then(res => res.template);
        // @ts-expect-error
        return yeoman_env_1.default.run(template, Object.assign(Object.assign({}, flags), { domain, path: folder, prismic: this.prismic, existingRepo }));
    }
    async validateTemplate(template, options) {
        if (template && options.includes(template))
            return Promise.resolve(template);
        return inquirer.prompt({
            type: 'list',
            name: 'template',
            message: 'Template to use',
            choices: options,
        }).then(res => res.template);
    }
}
exports.default = New;
New.description = 'Create a project with a new Prismic repository.';
New.flags = {
    help: command_1.flags.help({ char: 'h' }),
    domain: command_1.flags.string({
        char: 'd',
        description: 'Name of the Prismic repository. For example: repo-name, becomes https://repo-name.prismic.io.',
        parse: (input) => input.toLowerCase().trim(),
    }),
    folder: command_1.flags.string({
        char: 'f',
        description: 'Name of the project folder.',
    }),
    template: command_1.flags.string({
        char: 't',
        description: 'Prismic template for the project.',
    }),
    force: command_1.flags.boolean({ description: 'Overwrite local files.' }),
    'skip-install': command_1.flags.boolean({
        description: 'Prevent running install command after generating project.',
        default: false,
    }),
    'existing-repo': command_1.flags.boolean({
        description: 'Connect to an existing Prismic repository.',
        default: false,
    }),
};
New.args = [];
