// prismic-cli/lib/commands/signup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const prismic_1 = require("../prismic");
const cli_ux_1 = require("cli-ux");
const server_1 = require("../utils/server");
const logDecoration_1 = require("../utils/logDecoration");
class Signup extends prismic_1.Command {
    async run() {
        const { flags } = this.parse(Signup);
        const { base, port, 'auth-url': authUrl } = flags;
        // ask confirmation
        const confirmationMessage = logDecoration_1.PRISMIC_LOG_HEADER + 'Press any key to open up the browser to signup or ' + logDecoration_1.LogDecorations.FgRed + 'q' + logDecoration_1.LogDecorations.Reset + ' to exist';
        const confirmationKey = await cli_ux_1.default.prompt(confirmationMessage, { type: 'single', required: false });
        if (confirmationKey === 'q' || confirmationKey === '\u0003')
            return Promise.resolve();
        return this.prismic.signUp(port, base, authUrl);
    }
}
exports.default = Signup;
Signup.description = 'Create a Prismic account.';
Signup.flags = {
    help: command_1.flags.help({ char: 'h' }),
    base: command_1.flags.string({
        name: 'base',
        hidden: true,
        description: 'Base URL to sign up with.',
        default: 'https://prismic.io',
    }),
    port: command_1.flags.integer({
        name: 'port',
        hidden: false,
        description: 'Port to start the local login server.',
        default: server_1.DEFAULT_PORT,
    }),
    'auth-url': command_1.flags.string({
        hidden: true,
        name: 'auth-url',
        description: 'URL to use when validating and refreshing sessions.',
    }),
};
