// prismic-cli/lib/commands/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
const prismic_1 = require("../prismic");
const yeoman_env_1 = require("../prismic/yeoman-env");
class Theme extends prismic_1.Command {
    async run() {
        const isAuthenticated = await this.prismic.isAuthenticated();
        if (!isAuthenticated) {
            await this.login();
        }
        const { flags, args } = this.parse(Theme);
        const existingRepo = flags['existing-repo'] || false;
        const domain = await this.validateDomain(flags.domain, existingRepo);
        const folder = await this.validateFolder(flags.folder, domain, flags.force);
        const theme = await this.validateTheme(flags['theme-url'] || args.source);
        yeoman_env_1.default.register(require.resolve('../generators/theme'), 'Theme');
        const { conf: configPath, documents: documentsPath, customTypes: customTypesPath } = flags, rest = tslib_1.__rest(flags
        // @ts-expect-error
        , ["conf", "documents", "customTypes"]);
        // @ts-expect-error
        return yeoman_env_1.default.run('Theme', Object.assign({ domain, source: theme, path: folder, prismic: this.prismic, configPath, existingRepo, documentsPath, customTypesPath }, rest));
    }
}
exports.default = Theme;
Theme.description = 'Create a project from a ZIP file or a GitHub repository URL and a new Prismic repository.';
Theme.flags = {
    help: command_1.flags.help({ char: 'h' }),
    domain: command_1.flags.string({
        char: 'd',
        description: 'Name of the new Prismic repository. For example, repo-name becomes https://repo-name.prismic.io.',
        parse: input => input.toLowerCase().trim(),
    }),
    folder: command_1.flags.string({
        char: 'f',
        description: 'Name of the project folder.',
    }),
    'theme-url': command_1.flags.string({
        char: 't',
        description: 'GitHub URL or path to the theme file.',
    }),
    conf: command_1.flags.string({
        char: 'c',
        description: 'Path to Prismic configuration file.',
        default: 'prismic-configuration.js',
    }),
    documents: command_1.flags.string({
        description: 'Path to the documents in the theme.',
        default: 'documents',
    }),
    customTypes: command_1.flags.string({
        description: 'Path to the Custom Types directory in the theme.',
        default: 'custom_types',
    }),
    force: command_1.flags.boolean({ description: 'Overwrite local files.' }),
    'skip-install': command_1.flags.boolean({
        description: 'Prevent running install command after generating project.',
        default: false,
    }),
    'existing-repo': command_1.flags.boolean({
        description: 'Connect to an existing Prismic repository.',
        default: false,
    }),
};
Theme.args = [{
        name: 'source',
        description: 'Path or URL to a ZIP file, or a GitHub repository for the theme.',
        required: false,
    }];
