// prismic-cli/lib/prismic/yeoman-env.js
"use strict";
/* eslint-disable unicorn/no-abusive-eslint-disable, @typescript-eslint/ban-ts-ignore */
Object.defineProperty(exports, "__esModule", { value: true });
exports.apps = exports.filterMetaFor = exports.meta = exports.names = exports.all = void 0;
const yeoman_environment_1 = require("yeoman-environment");
const env = yeoman_environment_1.createEnv();
exports.all = env.lookup({
    // @ts-ignore
    lookups: ['generators'],
    packagePatterns: ['generator-prismic-*'],
});
exports.names = env.getGeneratorNames(); // names of the generators
exports.meta = env.getGeneratorsMeta(); // local first logic
function filterMetaFor(generatorsMeta, subGeneratorName) {
    return Object.entries(generatorsMeta)
        .reduce((acc, [key, value]) => {
        if (key.endsWith(`:${subGeneratorName}`)) {
            return Object.assign(Object.assign({}, acc), { [key]: value });
        }
        return acc;
    }, {});
}
exports.filterMetaFor = filterMetaFor;
exports.apps = filterMetaFor(exports.meta, 'app');
exports.default = env;
