// @prismicio/prismic-generator-generator/app/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Generator = require("yeoman-generator");
const path = require("path");
const inquirer = require("inquirer");
class default_1 extends Generator {
    constructor(argv, opts) {
        super(argv, opts);
        this.pm = opts.pm;
        this.name = opts.name;
        this.language = opts.language;
        this.path = opts.path;
    }
    async initializing() {
        if (this.path)
            this.destinationRoot(this.path);
    }
    async prompting() {
        if (!this.name) {
            this.name = await inquirer.prompt([
                {
                    type: 'input',
                    name: 'name',
                    message: 'name of the generator',
                    transformer: value => `generator-prismic-${value}`,
                    validate: value => value ? true : 'required',
                },
            ]).then(res => res.name.trim());
        }
        if (!this.language) {
            this.language = await inquirer.prompt([
                {
                    type: 'list',
                    name: 'language',
                    default: 'javascript',
                    choices: [
                        { name: 'JavaScript', value: 'javascript' },
                        { name: 'TypeScript', value: 'typescript' },
                    ],
                    message: 'Language',
                },
            ]).then(res => res.language);
        }
        if (!this.pm) {
            this.pm = await inquirer.prompt([
                {
                    type: 'list',
                    name: 'pm',
                    message: 'package manager',
                    choices: [
                        {
                            name: 'Yarn',
                            value: 'yarn',
                        },
                        {
                            name: 'Npm',
                            value: 'npm',
                        },
                    ],
                },
            ]).then(res => res.pm);
        }
        if (!this.slicemachine) {
            this.slicemachine = await inquirer.prompt([{
                    type: 'confirm',
                    name: 'slicemachine',
                    message: 'Support SliceMachine',
                    default: true,
                }]).then(res => res.slicemachine);
        }
    }
    async configuring() {
        this.destinationRoot(`generator-prismic-${this.name}`);
    }
    async writing() {
        const template = path.join(this.language || 'javascript', '**');
        const opts = {
            name: this.name,
            packageName: `generator-prismic-${this.name}`,
            slicemachine: this.slicemachine,
        };
        this.fs.copyTpl(this.templatePath(template), this.destinationPath(), opts, {}, { globOptions: { dot: true } });
        if (!this.slicemachine) {
            this.deleteDestination(path.join('generators', 'slicemachine'));
            this.deleteDestination(path.join('generators', 'create-slice'));
            this.deleteDestination(path.join('generators', 'storybook'));
        }
    }
    async install() {
        if (this.pm === 'yarn') {
            return this.yarnInstall();
        }
        return this.npmInstall();
    }
}
exports.default = default_1;
