// prismic-cli/lib/utils/framework/project-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedProjects = exports.findProjectType = exports.ProjectType = void 0;
var ProjectType;
(function (ProjectType) {
    ProjectType["NUXT"] = "nuxt";
    ProjectType["NEXT"] = "next";
    ProjectType["VUE"] = "vue.js";
    ProjectType["REACT"] = "react";
    ProjectType["SVELTE"] = "svelte";
    ProjectType["GATSBY"] = "gatsby";
    ProjectType["NODE"] = "node";
})(ProjectType = exports.ProjectType || (exports.ProjectType = {}));
function findProjectType(strFramework) {
    switch (strFramework.toLowerCase()) {
        case ProjectType.NUXT: return ProjectType.NUXT;
        case ProjectType.NEXT: return ProjectType.NEXT;
        case ProjectType.VUE: return ProjectType.VUE;
        case ProjectType.REACT: return ProjectType.REACT;
        case ProjectType.SVELTE: return ProjectType.SVELTE;
        case ProjectType.GATSBY: return ProjectType.GATSBY;
        case ProjectType.NODE: return ProjectType.NODE;
        default: return null;
    }
}
exports.findProjectType = findProjectType;
/*
 * This has to be an array sorted in order of specificity/priority.
 * Reason: both REACT and WEBPACK_REACT have react as dependency,
 * therefore WEBPACK_REACT has to come first, as it's more specific.
 */
exports.supportedProjects = [
    {
        preset: ProjectType.NUXT,
        dependencies: ['nuxt'],
        matcherFunction: ({ dependencies }) => dependencies.some(Boolean),
    },
    {
        preset: ProjectType.NEXT,
        dependencies: ['next'],
        matcherFunction: ({ dependencies }) => dependencies.some(Boolean),
    },
    {
        preset: ProjectType.VUE,
        dependencies: ['vue'],
        matcherFunction: ({ dependencies }) => dependencies.some(Boolean),
    },
    {
        preset: ProjectType.REACT,
        dependencies: ['react'],
        matcherFunction: ({ dependencies }) => dependencies.every(Boolean),
    },
    {
        preset: ProjectType.SVELTE,
        dependencies: ['svelte'],
        matcherFunction: ({ dependencies }) => dependencies.every(Boolean),
    },
    {
        preset: ProjectType.GATSBY,
        dependencies: ['gatsby'],
        matcherFunction: ({ dependencies }) => dependencies.every(Boolean),
    },
    {
        preset: ProjectType.NODE,
        dependencies: ['express', 'koa'],
        matcherFunction: ({ dependencies }) => dependencies.every(Boolean),
    },
];
