// prismic-cli/lib/generators/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const prismic_yeoman_generator_1 = require("@prismicio/prismic-yeoman-generator");
const Framework = require("../utils/framework");
class PrismicTheme extends prismic_yeoman_generator_1.default {
    constructor(argv, opts) {
        super(argv, opts);
        this.source = opts.source;
        this.customTypesPath = opts.customTypesPath;
        this.documentsPath = opts.documentsPath;
        this.configPath = opts.configPath || 'prismic-configuration.js';
    }
    async initializing() {
        this.destinationRoot(this.path);
        const innerFolder = this.innerFolderFromGitRepo(this.source);
        return this.downloadAndExtractZipFrom(this.source, innerFolder);
    }
    async prompting() {
        this.pm = await this.promptForPackageManager();
    }
    async configuring() {
        const customTypes = this.readCustomTypesFrom(this.customTypesPath);
        const documents = this.readDocumentsFrom(this.documentsPath);
        const pkg = (() => {
            const pkgPath = this.destinationPath('package.json');
            if (this.fs.exists(pkgPath) === false) {
                return null;
            }
            return this.fs.readJSON(pkgPath);
        })();
        if (!pkg)
            console.error('NO PKG FOUND');
        const maybeFramework = pkg && Framework.detect(pkg);
        this.maybeCreatePrismicRepository({
            domain: this.domain,
            customTypes,
            signedDocuments: documents,
            framework: maybeFramework || 'other',
        }, this.existingRepo).then(res => {
            if (res.data)
                this.domain = res.data;
        });
    }
    async writing() {
        const location = this.destinationPath(this.configPath);
        let themeConfig = {};
        try {
            themeConfig = JSON.parse(this.readDestination('prismic-theme.json', { defaults: '{}' }));
        }
        catch (_a) {
            console.error("An invalid prismic-theme.json was found. The theme's configuration will be ignored.");
        }
        if (this.existsDestination(this.configPath)) {
            const oldConfig = this.readDestination(this.configPath);
            const oldRepoName = themeConfig.replaceRepositoryName || 'your-repo-name';
            const newConfig = oldConfig.replace(new RegExp(oldRepoName, 'g'), this.domain);
            this.writeDestination(this.configPath, newConfig);
        }
        else {
            const url = new URL(this.prismic.base);
            url.host = `${this.domain}.${url.host}`;
            url.pathname = '/api/v2';
            this.fs.writeJSON(location, { apiEndpoint: url.toString() });
        }
        if (this.existsDestination('prismic-theme.json')) {
            this.deleteDestination('prismic-theme.json');
        }
    }
    async install() {
        if (this.pm === 'yarn') {
            this.yarnInstall();
        }
        else {
            this.npmInstall();
        }
    }
}
exports.default = PrismicTheme;
