// generator-prismic-nuxt/generators/storybook/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const prismic_yeoman_generator_1 = require("@prismicio/prismic-yeoman-generator");
const add_get_stories_paths_1 = require("./add-get-stories-paths");
const { SM_FILE } = require('sm-commons/consts');
class StoryBookNuxt extends prismic_yeoman_generator_1.default {
    constructor(argv, opts) {
        super(argv, opts);
        // this.pm = this.pm || 'npm'
        if (this.destinationRoot().endsWith(this.path) === false) {
            this.destinationRoot(this.path);
        }
    }
    async prompting() {
        if (!this.pm)
            await this.promptForPackageManager();
    }
    async writing() {
        const pkJson = {
            scripts: {
                storybook: 'nuxt storybook',
                'build-storybook': 'nuxt storybook build',
            },
            devDependencies: {
                sass: '^1.35.1',
                '@storybook/vue': '6.1.21',
                'vue-loader': '15.9.6',
                '@nuxtjs/storybook': '3.3.1',
                'sass-loader': '^10.1.1',
                'slice-machine-ui': '^0.1.2',
            },
        };
        this.fs.extendJSON(this.destinationPath('package.json'), pkJson);
        const smJson = {
            storybook: 'http://localhost:3003',
        };
        this.fs.extendJSON(this.destinationPath(SM_FILE), smJson);
        const config = this.readDestination('nuxt.config.js');
        const updatedConfig = add_get_stories_paths_1.default(config);
        this.writeDestination('nuxt.config.js', updatedConfig);
    }
    async install() {
        if (this.pm === 'yarn') {
            return this.yarnInstall();
        }
        return this.npmInstall();
    }
}
exports.default = StoryBookNuxt;
