// prismic-cli/lib/utils/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startServerAndOpenBrowser = exports.Server = exports.Routes = exports.DEFAULT_PORT = void 0;
const cli_ux_1 = require("cli-ux");
const logDecoration_1 = require("./logDecoration");
const hapi = require("@hapi/hapi");
exports.DEFAULT_PORT = 5555;
function validatePayload(payload) {
    if (!payload)
        return null;
    if (!payload.email || !payload.cookies)
        return null;
    if (!(Array.isArray(payload.cookies)))
        return null;
    if (payload.cookies.some((c) => typeof c !== 'string'))
        return null;
    return payload;
}
const authenticationHandler = (server) => (cb) => {
    return async (request, h) => {
        try {
            const data = validatePayload(request.payload);
            if (!data) {
                cli_ux_1.default.action.stop('It seems the server didn\'t respond properly, please contact us.');
                return h.response('Error with cookies').code(400);
            }
            await cb(data);
            return h.response(data).code(200);
        }
        finally {
            server.stop({ timeout: 10000 });
        }
    };
};
exports.Routes = {
    authentication: (server) => (cb) => ({
        method: 'POST',
        path: '/',
        handler: authenticationHandler(server)(cb),
    }),
    notFound: {
        method: ['GET', 'POST'],
        path: '/{any*}',
        handler: (request, h) => {
            return h.response(`not found: [${request.method}]: ${request.url.toString()}`).code(404);
        },
    },
};
exports.Server = {
    build: (base, port, host) => {
        const server = hapi.server({
            port,
            host,
            routes: {
                cors: {
                    origin: [base],
                    headers: ['Origin', 'X-Requested-With', 'Content-Type', 'Accept'],
                },
            },
        });
        return server;
    },
};
async function startServerAndOpenBrowser(url, base, port, logAction, setCookies) {
    return new Promise(resolve => {
        async function callback(data) {
            await setCookies(data.cookies);
            cli_ux_1.default.action.stop(`Logged in as ${data.email}`);
            resolve();
        }
        const server = exports.Server.build(base, port, 'localhost');
        server.route([exports.Routes.authentication(server)(callback), exports.Routes.notFound]);
        server.start()
            .then(() => {
            cli_ux_1.default.log('\nOpening browser to ' + logDecoration_1.LogDecorations.Underscore + url + logDecoration_1.LogDecorations.Reset);
            cli_ux_1.default.action.start(logAction, 'Waiting for the browser response');
            cli_ux_1.default.open(url);
        });
    });
}
exports.startServerAndOpenBrowser = startServerAndOpenBrowser;
exports.default = startServerAndOpenBrowser;
